src/os/bluestore/BlueFS.cc

```diff
@@ -1391,7 +1393,15 @@ int BlueFS::_flush_and_sync_log(std::unique_lock<std::mutex>& l,
     int r = _allocate(log_writer->file->fnode.prefer_bdev,
 		      cct->_conf->bluefs_max_log_runway,
 		      &log_writer->file->fnode.extents);
+    if ((r == -ENOSPC) && _should_compact_log() && _try_num) {
+      // log_writer->file->fnode.extents may hint the end of the dev...
+      _compact_log_sync(); 
```

Comment:
ohh yes...could we do the `::encode(log_t, bl);`  before call `_compact_log_sync()`  and lock the log_t? 
as indicated by tracker: #20645, EVEN the BDEV_WAL was not exhausted... the wal log file hold the first 400M of first 1G Zone, and the log_write'file(ino=1) hold the last 300MB of last 1G zone, not sure this is normal? when the log_write want extend its runway, but it hint the end of the BDEV_WAL(5GB), so allocator cant fail to other device? expecting your correct, Sir